### PR TITLE
FileReadSchemaTransform

### PR DESCRIPTION
**Pull request by ahmedabu98**
Implement a SchemaTransform for file reads. Will support formats: Avro, JSON, Parquet, and reading individual lines.

Enables watching for new files by setting a poll interval.

### CODE REVIEW

File: sdks/java/io/common/src/test/java/org/apache/beam/sdk/io/common/SchemaAwareJavaBeans.java
Comment (damondouglas) [inline sdks/java/io/common/src/test/java/org/apache/beam/sdk/io/common/SchemaAwareJavaBeans.java:409]: I'll let you decide based on my comment and I'll LGTM either way.

What I like about AvroPrimitiveDataTypes is that it makes it easy to write tests against field types we know would be compatible with Avro encodings.  What I struggle with is:
1. The name implies a coupling of data encoding though I get it
2. Writing tests against AllPrimitiveDataTypes that fail is valuable information on how we expose FieldType
3. Dealing with annoying FieldTypes that aren't compatible with a Schema related use case is valuable information to figure out how a real user of Beam would have to deal with this when they don't have the ability to change the original class but do have the ability to change the Schema using, for example, https://beam.apache.org/releases/javadoc/current/org/apache/beam/sdk/schemas/transforms/DropFields.html.  I dealt with this problem when writing tests for CsvIO and it was valuable discovery.  Writing tests around this problem of incompatible fields provided for me a usability test not just does this code work test.

Another alternative would be just to get rid of the Byte field in AllPrimitiveDataTypes but that would require you to refactor the tests that break as a result.

This comment also applies to the other additions to SchemaAwareJavaBeans and I didn't comment on those individually as it would have been redundant.

Comment (ahmedabu98) [reply sdks/java/io/common/src/test/java/org/apache/beam/sdk/io/common/SchemaAwareJavaBeans.java:409]: Yeah the problem is Short (int16)  and Byte (a single Byte) are not types that are supported by Avro (see [Avro primitive types](https://docs.oracle.com/cd/E26161_02/html/GettingStartedGuide/avroschemas.html#:~:text=in%20this%20chapter.-,Primitive%20Data%20Types,-In%20the%20previous)) or Parquet (see [primitive types](https://parquet.apache.org/docs/file-format/types/)). So actually in AvroUtils we turn Short and Byte types into Beam Schema types to an Integer (int32).

Avro and Parquet do support a sequence of bytes though, so I've replaced Byte with ByteBuffer in `AvroPrimitiveDataTypes`.

>Another alternative would be just to get rid of the Byte field in AllPrimitiveDataTypes

If you agree, I'd be happy to remove Short and replace Byte with ByteBuffer in `AllPrimitiveDataTypes` and use that for everything instead (for both read and write).

Comment (damondouglas) [reply sdks/java/io/common/src/test/java/org/apache/beam/sdk/io/common/SchemaAwareJavaBeans.java:409]: If you need my help refactoring tests that break as a result of removing Short and Byte, please let me know and I can commit to this PR.

Comment (ahmedabu98) [reply sdks/java/io/common/src/test/java/org/apache/beam/sdk/io/common/SchemaAwareJavaBeans.java]: Tried this, looks like XML, JSON, CSV don't support bytes (as in a sequence of bytes), but do support a single byte type. Opposite to the case of Avro and Parquet, which support a byte sequence but not a single byte.

Comment (ahmedabu98) [reply sdks/java/io/common/src/test/java/org/apache/beam/sdk/io/common/SchemaAwareJavaBeans.java]: If you're not opposed, will proceed by not including any form of bytes in `AllPrimitiveDataTypes` since there is no common overlap. Instead, we can have two new classes, ByteType and ByteSequenceType that will test for byte, byte[]. I can create new tests for these in both file read and write transforms.

Comment (damondouglas) [reply sdks/java/io/common/src/test/java/org/apache/beam/sdk/io/common/SchemaAwareJavaBeans.java]: I like your idea about removing the troublesome byte types out.  Let me know if you need help refactoring tests.

**Comment by ahmedabu98** [comment]
Run Java_File-schema-transform_IO_Direct PreCommit

**Comment by ahmedabu98** [comment]
R: @damondouglas

**Comment by ahmedabu98** [comment]
R: @johnjcasey 
R: @Abacn

File: sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/utils/JsonUtils.java
Comment (johnjcasey) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/utils/JsonUtils.java:160]: Nit: I'd prefer not to include new @suppressWarnings blocks if possible

File: sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/utils/JsonUtils.java
Comment (johnjcasey) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/utils/JsonUtils.java:237]: what are the consequences of this?

Comment (ahmedabu98) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/utils/JsonUtils.java:237]: properties in a json schema do not maintain ordering. When we convert a Beam schema to a json schema, we add the fields to the required properties in order. So when converting from json schema back to Beam schema, the ordering is kept in required properties. As for the additional (non-required) properties, we won't be able to guarantee the Beam schema order. 

Unlike json schemas, Beam schemas care about order. So one json schema with non-required properties can output more than one beam schema, and for Beam standards they would be equivalent but not equal. This makes a best effort attempt to make a given Beam schema survive a roundtrip of converting to json schema and back.

**Comment by johnjcasey** [comment]
run java precommit

File: sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/utils/JsonUtils.java
Comment (johnjcasey) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/utils/JsonUtils.java:235]: What is the consequences of failing the best efforts? can we detect the failure, and do something about it?

**Comment by ahmedabu98** [comment]
Run Spotless PreCommit

**Comment by ahmedabu98** [comment]
Run Spotless PreCommit

**Comment by ahmedabu98** [comment]
Run Java PreCommit

File: sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/utils/JsonUtils.java
Comment (ahmedabu98) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/utils/JsonUtils.java:235]: Sorry I worded it wrong. This **does** guarantee that a given Beam schema survives a roundtrip of converting to JSON schema and back. The explanation is because when we convert from Beam schema to Json schema, we specify all fields as required. So when converting from Json schema to Beam schema, the order is preserved in the Json `required` array. The takeaway is that a Json schema generated from a Beam schema will have all of its properties as required.

However, it is not guaranteed that a given Json schema **with non-required properties** will always output the same Beam Schema. It will be the same fields, but order is not predictable because non-required properties are accessed by a map.
In the latter case, the failure will show up when working with Beam Schemas and Rows in a use-case where Schema equality matter. The Schemas will be [equivalent](https://github.com/apache/beam/blob/3609766cd56c0b78e62757dc3fe848fd8ce01c1b/sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/Schema.java#L432-L435), but may not always be equal.

Comment (johnjcasey) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/utils/JsonUtils.java:235]: got it, LGTM

File: sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/FileReadSchemaTransformConfiguration.java
Comment (damondouglas) [inline sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/FileReadSchemaTransformConfiguration.java:32]: Is there a way to apply nullness warnings suppression at the methods that are giving the problem? It's more code but alternatively do you like the idea of using Optional?  In the configuration I don't imagine a performance hit as it is evaluated at pipeline construction time.

File: sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/FileReadSchemaTransformConfiguration.java
Comment (damondouglas) [inline sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/FileReadSchemaTransformConfiguration.java:50]: I'm struggling with the need for someone to look into this code to determine that schema is required for all of the formats except for line.  Neither of the following options are 100% ideal and I trust your judgement but wanted to relay my thoughts.

1) Remove LineReadSchemaTransformFormatProvider and the Nullable annotation on the getSchema method
2) Remove the Nullable annotation on the getSchema method and provide/document a Schema for LineReadSchemaTransformFormatProvider

What's missing from option 1 and 2 is dealing with a non-null empty string. I'd have to test this on my end but I wonder if Schema is one of the rare non-primitive types that we can get away with.

File: sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/FileReadSchemaTransformConfiguration.java
Comment (damondouglas) [inline sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/FileReadSchemaTransformConfiguration.java:100]: Should this be a Long so that it is easy to pass into [Duration](https://www.joda.org/joda-time/apidocs/org/joda/time/Duration.html#Duration-long-)?

File: sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/FileReadSchemaTransformProvider.java
Comment (damondouglas) [inline sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/FileReadSchemaTransformProvider.java:105]: I had a hard time understanding this. Is this because we have the LineReadSchemaTransformFormatProvider?

**Review by damondouglas — CHANGES_REQUESTED** [review]
I haven't finished reviewing but these are some initial thoughts.

File: sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/FileReadSchemaTransformConfiguration.java
Comment (damondouglas) [reply sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/FileReadSchemaTransformConfiguration.java:50]: I did an experiment that might help here. The experiment answers "Can we use a Schema as a field type in our SchemaTransform configurations?"  If yes, the implications of the answer in FileReadSchemaTransformConfiguration's context would be:

```
abstract class FileReadSchemaTransformConfiguration {
    abstract Schema getSchema();
}
```

So far it looks like yes.  Here's what I did:

#### 1. Define a simple AutoValue class

<details>
<summary>AutoValue Class with Schema field</summary>

```
@DefaultSchema(AutoValueSchema.class)
@AutoValue
public abstract class ExampleModel {

  public static Builder builder() {
    return new AutoValue_ExampleModel.Builder();
  }

  public abstract String getAString();

  public abstract Schema getASchema();

  @AutoValue.Builder
  public abstract static class Builder {

    public abstract Builder setAString(String value);

    public abstract Builder setASchema(Schema value);

    public abstract ExampleModel build();
  }
}
```
</details>

#### 2. Generate to/from Row functions from schema provider

<details>
<summary>SchemaProvider boilerplate</summary>

```
private static final AutoValueSchema SCHEMA_PROVIDER = new AutoValueSchema();
private static final TypeDescriptor<ExampleModel> TD = TypeDescriptor.of(ExampleModel.class);
private static final SerializableFunction<ExampleModel, Row> TO_ROW_FN = SCHEMA_PROVIDER
      .toRowFunction(TD);
private static final SerializableFunction<Row, ExampleModel> FROM_ROW_FN = SCHEMA_PROVIDER
      .fromRowFunction(TD);
```
</details>

#### 3. Instantiate the AutoValue class, setting the Schema field

<details>
<summary>Create schema then instantiate ExampleModel</summary>

```
Schema schema = Schema.of(
    Field.of("anInteger", FieldType.INT32),
    Field.of("aBoolean", FieldType.BOOLEAN),
    Field.of("something", FieldType.STRING)
);
ExampleModel model = ExampleModel.builder()
    .setAString("hello")
    .setASchema(schema)
    .build();
```

</details>

#### 4. Convert ExampleModel to a Row

This simulates what the SchemaProvider would receive as the configuration representation.

```
Row row = TO_ROW_FN.apply(model);
```

#### 5. Convert back to an ExampleModel from the Row

```
ExampleModel backAgain = FROM_ROW_FN.apply(row);
```

#### 6. Finally check whether the Schema fields were preserved during the conversions

<details>
<summary>Call getSchema() and run a bunch of checks</summary>

```
Schema backAgainASchema = backAgain.getASchema();
checkNotNull(backAgainASchema);
checkState(backAgainASchema.hasField("anInteger"));
checkState(backAgainASchema.hasField("aBoolean"));
checkState(backAgainASchema.hasField("something"));
checkState(backAgainASchema.getField("anInteger").getType().equals(FieldType.INT32));
checkState(backAgainASchema.getField("aBoolean").getType().equals(FieldType.BOOLEAN));
checkState(backAgainASchema.getField("something").getType().equals(FieldType.STRING));
```
</details>

File: sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/utils/JsonUtils.java
Comment (damondouglas) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/utils/JsonUtils.java:169]: Does everit correctly parse the JSON integer value to the Java type either INT32 or INT64?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/utils/JsonUtils.java
Comment (damondouglas) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/utils/JsonUtils.java:173]: Same as https://github.com/apache/beam/pull/25927/files?file-filters%5B%5D=.java&show-viewed-files=false#r1167333670

File: sdks/java/io/file-schema-transform/src/test/java/org/apache/beam/sdk/io/fileschematransform/AvroReadSchemaTransformFormatProviderTest.java
Comment (damondouglas) [inline sdks/java/io/file-schema-transform/src/test/java/org/apache/beam/sdk/io/fileschematransform/AvroReadSchemaTransformFormatProviderTest.java:89]: Should we test whether files written using FileWriteSchemaTransformProvider read from FileReadSchemaTransformProvider?

File: sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/FileReadSchemaTransformConfiguration.java
Comment (ahmedabu98) [reply sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/FileReadSchemaTransformConfiguration.java:50]: Thanks for the analysis Damon! How would it look like for Python to send a Schema object? I'd prefer using a String for now to keep the cross-language use-case simple. 

> I'm struggling with the need for someone to look into this code to determine that schema is required for all of the formats except for line

Would it help to improve the error message? I can include all the formats that do require a schema.

File: sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/FileReadSchemaTransformConfiguration.java
Comment (ahmedabu98) [reply sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/FileReadSchemaTransformConfiguration.java:100]: Was trying to keep cross-language use case in mind, but looks like Python ints are translated to INT64 types: https://github.com/apache/beam/blob/f2f2fb83a21c231d741008d5fc6fdda7c69bb786/sdks/python/apache_beam/typehints/schemas.py#L28
Thanks for the catch, will change it to Long

File: sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/FileReadSchemaTransformProvider.java
Comment (ahmedabu98) [reply sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/FileReadSchemaTransformProvider.java:105]: No this is for all the read schematransforms. Basically there are two use-cases:

1. read files that match **one** filepattern (ie. use a configuration field)
2. read files that match **multiple** filepatterns (ie. use an input PCollection<String> of filepattern elements)

File: sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/utils/JsonUtils.java
Comment (ahmedabu98) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/utils/JsonUtils.java:169]: Hmmm I'm not sure I understand. This library is for building json schemas, so it wouldn't be parsing any values. As far as these lines go, we're only concerned with whether the number is an integer or not.

File: sdks/java/io/file-schema-transform/src/test/java/org/apache/beam/sdk/io/fileschematransform/AvroReadSchemaTransformFormatProviderTest.java
Comment (ahmedabu98) [reply sdks/java/io/file-schema-transform/src/test/java/org/apache/beam/sdk/io/fileschematransform/AvroReadSchemaTransformFormatProviderTest.java:89]: Yeah probably worth doing

Comment (ahmedabu98) [reply sdks/java/io/file-schema-transform/src/test/java/org/apache/beam/sdk/io/fileschematransform/AvroReadSchemaTransformFormatProviderTest.java:89]: Added the tests

**Comment by damondouglas** [comment]
Run Java PreCommit

File: sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/FileReadSchemaTransformProvider.java
Comment (damondouglas) [reply sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/FileReadSchemaTransformProvider.java:105]: It makes sense now reading the Javadoc for the interface.

File: sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/FileReadSchemaTransformConfiguration.java
Comment (damondouglas) [reply sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/FileReadSchemaTransformConfiguration.java]: As we talked about, let's keep the String. In the future I would like to see if an actual Schema type would work and still be easy from a user perspective.

File: sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/FileReadSchemaTransformConfiguration.java
Comment (damondouglas) [reply sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/FileReadSchemaTransformConfiguration.java:32]: I git cloned this PR and tried these changes and was able to get a successful `./gradlew :sdks:java:io:file-schema-transform:check`. Would you consider to incorporating these changes?  I also provided comments at the lines of code what changes I was able to make to remove the null warnings without using the SuppressWarnings annotation.  I agree that we are applying `checkArgument` or `checkState` and why doesn't the checker pass kinds of questions. However, I hope that the following list could expedite the refactoring of this PR. (Note: Optional.isPresent seems to do the trick)

1. Remove all SupressWarnings annotations
2. Add a package private `getSafeSchema` method:

```
// FYI these are the imports:
// import java.util.Optional;
// import org.checkerframework.checker.nullness.qual.NonNull;

 @NonNull
  String getSafeSchema() {
    Optional<String> schema = Optional.ofNullable(getSchema());
    checkState(schema.isPresent() && !schema.get().isEmpty());
    return schema.get();
  }
```

3. Add a package private `getSafeFilepattern()`:

```
// FYI these are the imports:
// import java.util.Optional;
// import org.checkerframework.checker.nullness.qual.NonNull;
  @NonNull
  String getSafeFilepattern() {
    Optional<String> value = Optional.ofNullable(getFilepattern());
    checkState(value.isPresent() && !value.get().isEmpty());
    return value.get();
  }
```

File: sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/FileReadSchemaTransformProvider.java
Comment (damondouglas) [inline sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/FileReadSchemaTransformProvider.java:134]: 1) In the context of this [comment](https://github.com/apache/beam/pull/25927#discussion_r1174141095), this could change to `.via((row) -> Objects.requireNonNull(row.getString("filepattern"))))`

2) Instead of hardcoding things like "filepattern", could we instead have these constants just below the class declaration as `static final String ...`?

3) Should we check that the incoming row's Schema adhere's to having the `FieldType.STRING` field with the name "filepattern" and tests to show that meaningful errors display to the user about this expectation?

File: sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/FileReadSchemaTransformProvider.java
Comment (damondouglas) [inline sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/FileReadSchemaTransformProvider.java:178]: Instead of this method, may we consider an additional configuration field and throwing an error when the user provides both the Schema and the Schema file source property?  It took me some time to realize that the configuration's "Schema" property could be either a Schema or a file path.  Having a more declared intent by the user from where the Schema derives would simplify the code.

File: sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/FileReadSchemaTransformProvider.java
Comment (damondouglas) [inline sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/FileReadSchemaTransformProvider.java:173]: Instead of these two conditionals, in addition to the conditionals above as well as the conditional from which this method is called, may we consider a `buildMatchTransform` method that performs these checks and returns a `PTransform<PCollection<String>, PCollection<MatchResult.Metadata>>`?  The conditionals could be flattened making it easier to troubleshoot and test.  And, on the subject of testing either this method or `buildMatchTransform` pattern, may we see more test coverage?  Finally, does the method need to be public and not package private?

File: sdks/java/io/file-schema-transform/src/test/java/org/apache/beam/sdk/io/fileschematransform/FileReadSchemaTransformFormatProviderTest.java
Comment (damondouglas) [inline sdks/java/io/file-schema-transform/src/test/java/org/apache/beam/sdk/io/fileschematransform/FileReadSchemaTransformFormatProviderTest.java:157]: Could we see a test for the doublyNestedDataTypesRepeatRows?

**Review by damondouglas — CHANGES_REQUESTED** [review]

File: sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/FileReadSchemaTransformConfiguration.java
Comment (ahmedabu98) [reply sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/FileReadSchemaTransformConfiguration.java:32]: Sorry this suggestion slipped my mind last review. Will implement this, thanks for the examples

File: sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/FileReadSchemaTransformProvider.java
Comment (ahmedabu98) [reply sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/FileReadSchemaTransformProvider.java:181]: I prefer keeping one field that does both, I think that would be simpler from the user's perspective. One field that can handle both use cases.

Maybe the documentation for Schema isn't clear enough? I'll  try rewording it

File: sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/FileReadSchemaTransformProvider.java
Comment (ahmedabu98) [reply sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/FileReadSchemaTransformProvider.java]: Refactored into a buildMatchTransform as requested. I'm having trouble figuring out how to test this though as Match/MatchAll don't implement an equals method that would allow me to compare the generated match transform with an expected transform. Their internal MatchConfiguration objects are private attributes too.

**Review by damondouglas — APPROVED** [review]

**Pull request merged**
